In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from ast import literal_eval

import warnings
warnings.filterwarnings('ignore')

In [2]:
rawg_data = pd.read_csv('01_rawg_clean.csv', parse_dates = ['released'])

rawg_data.head()

slug                         name   released  \
0          grand-theft-auto-v           Grand Theft Auto V 2013-09-17   
1                    portal-2                     Portal 2 2011-04-18   
2     the-witcher-3-wild-hunt     The Witcher 3: Wild Hunt 2015-05-18   
3                 tomb-raider           Tomb Raider (2013) 2013-03-05   
4  the-elder-scrolls-v-skyrim  The Elder Scrolls V: Skyrim 2011-11-11   

   metacritic  suggestions_count  \
0        97.0                416   
1        95.0                582   
2        92.0                678   
3        86.0                664   
4        94.0                621   

                                           platforms  \
0  ['PC', 'Xbox Series S/X', 'PlayStation 5', 'Pl...   
1  ['Xbox One', 'PlayStation 3', 'PC', 'Xbox 360'...   
2  ['PC', 'Xbox One', 'Nintendo Switch', 'PlaySta...   
3  ['PC', 'PlayStation 4', 'PlayStation 3', 'Xbox...   
4  ['PC', 'PlayStation 3', 'Xbox 360', 'Nintendo ...   

                           genres  \
0         ['Action', 'Adventure']   
1           ['Shooter', 'Puzzle']   
2  ['Action', 'Adventure', 'RPG']   
3         ['Action', 'Adventure']   
4               ['Action', 'RPG']   

                                              stores  \
0  ['Epic Games', 'PlayStation Store', 'Xbox Stor...   
1  ['Xbox Store', 'Xbox 360 Store', 'PlayStation ...   
2  ['GOG', 'Xbox Store', 'Steam', 'PlayStation St...   
3  ['App Store', 'Google Play', 'PlayStation Stor...   
4  ['Xbox 360 Store', 'Nintendo Store', 'Steam', ...   

                                                tags   esrb_rating  
0  ['Singleplayer', 'Steam Achievements', 'Multip...        Mature  
1  ['Singleplayer', 'Steam Achievements', 'Multip...  Everyone 10+  
2  ['Singleplayer', 'Atmospheric', 'Full controll...        Mature  
3  ['Singleplayer', 'Multiplayer', 'Atmospheric',...        Mature  
4  ['Singleplayer', 'Steam Achievements', 'steam-...        Mature

# Unpacking columns

We want to unpack values inside the list-like columns, specially `platforms` (we can leave the rest for later, as we can directly do it on the feature engineering process).

The idea is to copy the raw for every platform listed in the column so that it looks like the following:

| slug | name | released | metacritic | suggestions_count | platforms | (...) |
| ---- | ---- | ---- | ---- | ---- | ---- | ---- |
| grand-theft-auto-v | Grand Theft Auto V | 2013-09-17 | 97.0 |	416 | PC | (...) |
| grand-theft-auto-v | Grand Theft Auto V | 2013-09-17 | 97.0 |	416 | Xbox Series S/X | (...) |
| grand-theft-auto-v | Grand Theft Auto V | 2013-09-17 | 97.0 |	416 | PlayStation 5 | (...) |


## But first...

CSV files has the problem that saves lists as str, so we have to convert them back to be able to access the data inside them. To do so, we will apply the `literal_eval` function from the `ast` library.

After the first cleaning, there are rows in the list-like columns which are not lists, as the list was a single-element list. We will not apply `literal_eval` on them as they raise an error.

In [3]:
rawg_data['platforms'] = rawg_data['platforms'].apply(lambda x: literal_eval(x) if pd.notnull(x) and ('[' in x) else x)
rawg_data['genres'] = rawg_data['genres'].apply(lambda x: literal_eval(x) if pd.notnull(x) and ('[' in x) else x)
rawg_data['stores'] = rawg_data['stores'].apply(lambda x: literal_eval(x) if pd.notnull(x) and ('[' in x) else x)
rawg_data['tags'] = rawg_data['tags'].apply(lambda x: literal_eval(x) if pd.notnull(x) and ('[' in x) else x)

---

## NaN values

Now that lists are lists again, let's look at the `NaN` values in this dataset to study them.

In [4]:
rawg_data.isna().sum()

slug                      0
name                      0
released              24608
metacritic           516685
suggestions_count         0
platforms              4077
genres               116156
stores                30070
tags                  62068
esrb_rating          463781
dtype: int64

---
### Platforms' NaNs


On a first glance, we may think thet it would be a good idea to remove them in the `platforms` column, as not having a platform means that the game won't be able to be merged into the VCG dataset.

Let's see what kind of games have `NaN` in the `platforms` column.

In [5]:
rawg_data[rawg_data['platforms'].isna()]

slug                   name   released  \
5211            half-life-2-downfall  Half-Life 2: Downfall 2017-04-17   
6967              minerva-metastasis    MINERVA: Metastasis 2005-09-02   
15281                      witchfire              Witchfire        NaT   
19904                     spacewar-2              Spacewar! 1962-01-01   
21229                  urban-legends          Urban Legends 2015-03-02   
...                              ...                    ...        ...   
510641  le-yuan-sheng-huo-hitsuzicun               楽園生活ひつじ村 2014-01-01   
510649                circle-of-mana         Circle of Mana 2013-03-05   
510769                       antraxx                Antraxx        NaT   
510771              the-pyramid-gate       The Pyramid Gate 2014-05-06   
510915                   miner-willy            Miner Willy        NaT   

        metacritic  suggestions_count platforms             genres   stores  \
5211           NaN                311       NaN  [Action, Shooter]    Steam   
6967           NaN                216       NaN            Shooter    Steam   
15281          NaN                185       NaN            Shooter      NaN   
19904          NaN                 37       NaN                NaN      NaN   
21229          NaN                201       NaN             Puzzle  itch.io   
...            ...                ...       ...                ...      ...   
510641         NaN                386       NaN                NaN      NaN   
510649         NaN                213       NaN                NaN      NaN   
510769         NaN                359       NaN            Shooter      NaN   
510771         NaN                 42       NaN          Adventure      NaN   
510915         NaN                  0       NaN                NaN      NaN   

                                                     tags esrb_rating  
5211                                       [Mod, destroy]         NaN  
6967                                                  Mod         NaN  
15281                                                 NaN         NaN  
19904                                     [Space, combat]         NaN  
21229                      [Space, puzzles, gun, Gravity]         NaN  
...                                                   ...         ...  
510641                                       Free to Play         NaN  
510649                     [battle, sword, balance, tree]         NaN  
510769        [Character Customization, Isometric, Mechs]         NaN  
510771  [Exploration, Pixel Graphics, Psychedelic, Abs...         NaN  
510915                                              night         NaN  

[4077 rows x 10 columns]

What we can see here is that when a game has a `NaN` on the `platforms` column, some other list-like columns can also have `NaN` on them.

However, if we look at, for example, the `stores` column more in detail...

In [6]:
rawg_data[rawg_data['platforms'].isna()]['stores'].value_counts()

itch.io              3273
PlayStation Store      34
Google Play             7
Steam                   4
Xbox Store              1
Name: stores, dtype: int64

We see that there is some portential problems there:

- A game that can be bought in steam should have `PC` on the `Platforms` column.

- The same happens to `Playstation Store` and `Xbox Store`. The problem here is that they are DLCs as we can see below, and they already have their counterpart for consoles on the dataset.

In [7]:
nan_plat = rawg_data[rawg_data['platforms'].isna()]
ps_store = rawg_data['stores'] == 'PlayStation Store'

nan_plat[ps_store]

slug  \
54769      werewolf-the-apocalypse-earthblood-ps4-and-ps5   
55152      wrc-9-fia-world-rally-championship-ps4-and-ps5   
55154                           borderlands-3-ps4-and-ps5   
71297                          dreaming-sarah-ps4-and-ps5   
73545                    yakuza-like-a-dragon-ps4-and-ps5   
75403                            thunderflash-ps4-and-ps5   
80966                anodyne-2-return-to-dust-ps4-and-ps5   
83968                         ultragoodness-2-ps4-and-ps5   
85228                       nioh-2-remastered-ps5-upgrade   
91736   atelier-ryza-2-lost-legends-and-the-secret-fai...   
94677                                  ride-4-ps4-and-ps5   
95710                                 fifa-21-ps4-and-ps5   
99574                               sparkle-2-ps4-and-ps5   
106928                       freddy-spaghetti-ps4-and-ps5   
106929    bridge-constructor-the-walking-dead-ps4-and-ps5   
110568                              destiny-2-ps4-and-ps5   
113582                           worms-rumble-ps4-and-ps5   
113585                 immortals-fenyx-rising-ps4-and-ps5   
113587                          madden-nfl-21-ps4-and-ps5   
116043                    concept-destruction-ps4-and-ps5   
116044                            tesla-force-ps4-and-ps5   
117433                             poker-club-ps4-and-ps5   
118787                sackboy-a-big-adventure-ps4-and-ps5   
118788       marvels-spider-man-miles-morales-ps4-and-ps5   
118791              mortal-kombat-11-ultimate-ps4-and-ps5   
118792                                  jydge-ps4-and-ps5   
121681                      watch-dogs-legion-ps4-and-ps5   
121682                           undead-horde-ps4-and-ps5   
121683                           the-pathless-ps4-and-ps5   
121684                            no-mans-sky-ps4-and-ps5   
121685                               maneater-ps4-and-ps5   
121686                           king-oddball-ps4-and-ps5   
121687                                 dirt-5-ps4-and-ps5   
121688      assassins-creed-valhalla-ultimate-ps4-and-ps5   

                                                     name   released  \
54769   Werewolf: The Apocalypse – Earthblood PS4 and PS5 2021-02-04   
55152      WRC 9 FIA World Rally Championship PS4 and PS5 2020-11-12   
55154                           Borderlands 3 PS4 and PS5 2020-11-12   
71297                          Dreaming Sarah PS4 and PS5 2021-03-05   
73545                   Yakuza: Like a Dragon PS4 and PS5 2021-03-02   
75403                            Thunderflash PS4 and PS5 2021-02-26   
80966               Anodyne 2: Return to Dust PS4 and PS5 2021-02-18   
83968                         UltraGoodness 2 PS4 and PS5 2021-02-09   
85228                     Nioh 2 Remastered (PS5 Upgrade) 2021-02-05   
91736   Atelier Ryza 2: Lost Legends and the Secret Fa... 2021-01-26   
94677                                  RIDE 4 PS4 and PS5 2021-01-21   
95710                                 FIFA 21 PS4 and PS5 2020-12-02   
99574                               Sparkle 2 PS4 and PS5 2020-12-21   
106928                       Freddy Spaghetti PS4 and PS5 2020-12-15   
106929  Bridge Constructor: The Walking Dead - PS4 and... 2020-12-15   
110568                              Destiny 2 PS4 and PS5 2020-12-08   
113582                           Worms Rumble PS4 and PS5 2020-12-01   
113585                 Immortals Fenyx Rising PS4 and PS5 2020-12-03   
113587                          Madden NFL 21 PS4 and PS5 2020-12-03   
116043                    Concept Destruction PS4 and PS5 2020-11-12   
116044                            Tesla Force PS4 and PS5 2020-11-24   
117433                             Poker Club PS4 and PS5 2020-11-20   
118787               Sackboy: A Big Adventure PS4 and PS5 2020-11-12   
118788     Marvel's Spider-Man: Miles Morales PS4 and PS5 2020-11-12   
118791              Mortal Kombat 11 Ultimate PS4 and PS5 2020-11-17   
118792                                  JYDGE

In [8]:
rawg_data[rawg_data['name'].str.contains('Assassin\'s Creed Valhalla')]

slug  \
954                          assassins-creed-valhalla   
121688  assassins-creed-valhalla-ultimate-ps4-and-ps5   

                                                  name   released  metacritic  \
954                          Assassin's Creed Valhalla 2020-11-10        83.0   
121688  Assassin's Creed Valhalla Ultimate PS4 and PS5 2020-11-12         NaN   

        suggestions_count                                          platforms  \
954                   449  [PlayStation 5, Xbox One, PC, Xbox Series S/X,...   
121688                512                                                NaN   

                          genres                                       stores  \
954     [Action, Adventure, RPG]  [PlayStation Store, Xbox Store, Epic Games]   
121688                       RPG                            PlayStation Store   

                                                     tags esrb_rating  
954                                    [Fantasy, vikings]      Mature  
121688  [Assassin, character, Epic, console, collectio...         NaN

In [9]:
rawg_data[rawg_data['name'].str.contains('Borderlands 3')]

slug  \
495                                        borderlands-3   
22381  borderlands-3-moxxis-heist-of-the-handsome-jac...   
55154                          borderlands-3-ps4-and-ps5   

                                                    name   released  \
495                                        Borderlands 3 2019-09-13   
22381  Borderlands 3: Moxxi’s Heist of the Handsome J... 2019-12-19   
55154                          Borderlands 3 PS4 and PS5 2020-11-12   

       metacritic  suggestions_count  \
495          83.0                693   
22381        78.0                373   
55154         NaN                275   

                                               platforms  \
495    [PC, Xbox Series S/X, PlayStation 5, PlayStati...   
22381                      [Xbox One, PlayStation 4, PC]   
55154                                                NaN   

                                  genres  \
495    [Action, Shooter, Adventure, RPG]   
22381       [Action, Shooter, Adventure]   
55154                  [Action, Shooter]   

                                                  stores  \
495    [Steam, PlayStation Store, Xbox Store, Epic Ga...   
22381        [PlayStation Store, Xbox Store, Epic Games]   
55154                                  PlayStation Store   

                                                    tags esrb_rating  
495    [Singleplayer, Steam Achievements, Multiplayer...      Mature  
22381                                                NaN      Mature  
55154  [Multiplayer, online, friends, console, skill,...         NaN

---
Let's see if we can gather more insight on the `Steam` department.

In [10]:
steam = rawg_data['stores'] == 'Steam'

nan_plat[steam]

slug                   name   released  metacritic  \
5211   half-life-2-downfall  Half-Life 2: Downfall 2017-04-17         NaN   
6967     minerva-metastasis    MINERVA: Metastasis 2005-09-02         NaN   
32216      endless-dungeons        Endless Dungeon        NaT         NaN   
54669  zenith-the-last-city  Zenith: The Last City        NaT         NaN   

       suggestions_count platforms                    genres stores  \
5211                 311       NaN         [Action, Shooter]  Steam   
6967                 216       NaN                   Shooter  Steam   
32216                326       NaN                    Action  Steam   
54669                260       NaN  [Action, Adventure, RPG]  Steam   

                 tags esrb_rating  
5211   [Mod, destroy]         NaN  
6967              Mod         NaN  
32216             NaN         NaN  
54669             NaN         NaN

What we see is that 2 of them are MODs (= free content) and the other 2 are games **not released yet**. Thus, we can remove all these data without losing information. as we do not want mobile-exclusive games and we will not consider itch.io for the moment, as the VGC dataset does not contain data from there, thus no sales data.

---

## Dropping rows

Let's then proceed with the removal of rows having the one or more of the next characteristics:

- `platforms` is `NaN`

- Mobile-only games, thus `platforms` only contain whether `Android` or `iOS`

- `stores` has only `itch.io`


In [11]:
plat_nan = list(rawg_data[rawg_data['platforms'].isna()].index)
android = list(rawg_data[rawg_data['platforms']=='Android'].index)
ios = list(rawg_data[rawg_data['platforms']=='iOS'].index)

plat_to_drop = plat_nan + android + ios

rawg_data = rawg_data.drop(index=plat_to_drop)\
                     .reset_index(drop=True)

rawg_data.shape[0]

439480

In [12]:
itchio = list(rawg_data[rawg_data['stores']=='itch.io'].index)

rawg_data = rawg_data.drop(index=itchio)\
                     .reset_index(drop=True)

rawg_data.shape[0]

94591

We have drastically reduced the dimension of the dataset to almost 20% of the original size.

Let's check how many `NaN` do we have:

In [13]:
rawg_data.isna().sum()

slug                     0
name                     0
released             13657
metacritic           89673
suggestions_count        0
platforms                0
genres               10253
stores               27811
tags                 19760
esrb_rating          79698
dtype: int64

In [14]:
nan_release = list(rawg_data[rawg_data['released'].isna()].index)

rawg_data = rawg_data.drop(index = nan_release)\
                     .reset_index(drop=True)

rawg_data

slug                         name   released  \
0              grand-theft-auto-v           Grand Theft Auto V 2013-09-17   
1                        portal-2                     Portal 2 2011-04-18   
2         the-witcher-3-wild-hunt     The Witcher 3: Wild Hunt 2015-05-18   
3                     tomb-raider           Tomb Raider (2013) 2013-03-05   
4      the-elder-scrolls-v-skyrim  The Elder Scrolls V: Skyrim 2011-11-11   
...                           ...                          ...        ...   
80929                crumble-zone                 Crumble Zone 2012-11-29   
80930  delta-strike-first-assault  Delta Strike: First Assault 2016-05-31   
80931          siege-hero-wizards           Siege Hero Wizards 2013-09-05   
80932                velocispider                 Velocispider 2011-06-01   
80933            kitten-sanctuary             Kitten Sanctuary 2009-03-13   

       metacritic  suggestions_count  \
0            97.0                416   
1            95.0                582   
2            92.0                678   
3            86.0                664   
4            94.0                621   
...           ...                ...   
80929         NaN                 95   
80930         NaN                356   
80931         NaN                193   
80932         NaN                 90   
80933         NaN                365   

                                               platforms  \
0      [PC, Xbox Series S/X, PlayStation 5, PlayStati...   
1      [Xbox One, PlayStation 3, PC, Xbox 360, Linux,...   
2         [PC, Xbox One, Nintendo Switch, PlayStation 4]   
3      [PC, PlayStation 4, PlayStation 3, Xbox 360, X...   
4         [PC, PlayStation 3, Xbox 360, Nintendo Switch]   
...                                                  ...   
80929                                     [Android, iOS]   
80930                                            PS Vita   
80931                                     [iOS, Android]   
80932                                     [iOS, Android]   
80933                                          [iOS, PC]   

                         genres  \
0           [Action, Adventure]   
1             [Shooter, Puzzle]   
2      [Action, Adventure, RPG]   
3           [Action, Adventure]   
4                 [Action, RPG]   
...                         ...   
80929          [Action, Arcade]   
80930                    Action   
80931          [Action, Puzzle]   
80932  [Action, Arcade, Casual]   
80933          [Family, Puzzle]   

                                                  stores  \
0      [Epic Games, PlayStation Store, Xbox Store, Xb...   
1      [Xbox Store, Xbox 360 Store, PlayStation Store...   
2            [GOG, Xbox Store, Steam, PlayStation Store]   
3      [App Store, Google Play, PlayStation Store, St...   
4      [Xbox 360 Store, Nintendo Store, Steam, PlaySt...   
...                                                  ...   
80929                           [App Store, Google Play]   
80930                                  PlayStation Store   
80931                           [App Store, Google Play]   
80932                           [App Store, Google Play]   
80933                                          App Store   

                                                    tags   esrb_rating  
0      [Singleplayer, Steam Achievements, Multiplayer...        Mature  
1      [Singleplayer, Steam Achievements, Multiplayer...  Everyone 10+  
2      [Singleplayer, Atmospheric, Full controller su...        Mature  
3      [Singleplayer, Multiplayer, Atmospheric, Full ...        Mature  
4      [Singleplayer, Steam Achievements, steam-tradi...        Mature  
...                                                  ...           ...  
80929  [Multiplayer, Space, Colorful, achievements, f...           NaN  
80930                     [combat, online, Tanks, drone]  Everyone 10+  
80931         [Physics, Cartoon, hero, Monsters, wizard]  Everyone 10+  
80932  [Retro, Robots, charact

---

Now we would want to unpack the rows to single-platform columns...but how do we accomplish that?

We could do the following:

- Create an auxiliar DataFrame `rawg_aux`

- Iterate over every row with `for idx, row in rawg_data.iterrows()`

- For every platform, append the columns to `rawg_aux`

\* **Ideally**, we would to do this step with a function being applied to the dataframe. 

However, we haven't come up with the idea to how to do it yet.

In [15]:
rawg_aux = pd.DataFrame(columns = rawg_data.columns)

In [16]:
for idx, row in rawg_data.iterrows():
    
    plats = row['platforms']
    
    if type(plats) != str: # We do not want to do this loop if the game only has one platform (thus, no list-like = string).
        for element in plats:      

            aux = row.copy()
            aux['platforms'] = element

            rawg_aux = rawg_aux.append(aux)
    
    else:
        
        aux = row.copy()
        aux['platforms'] = plats

        rawg_aux = rawg_aux.append(aux)
        
    if idx % 10000 == 0:
        print(idx)


rawg_aux = rawg_aux.rename(columns = {'platforms': 'platform',
                                      'esrb_rating': 'esrb'})

print('Finished!')

0
10000
20000
30000
40000
50000
60000
70000
80000
Finished!


In [17]:
rawg_data = rawg_aux.reset_index(drop=True)
rawg_data

slug                name   released  metacritic  \
0       grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   
1       grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   
2       grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   
3       grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   
4       grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   
...                    ...                 ...        ...         ...   
140611  siege-hero-wizards  Siege Hero Wizards 2013-09-05         NaN   
140612        velocispider        Velocispider 2011-06-01         NaN   
140613        velocispider        Velocispider 2011-06-01         NaN   
140614    kitten-sanctuary    Kitten Sanctuary 2009-03-13         NaN   
140615    kitten-sanctuary    Kitten Sanctuary 2009-03-13         NaN   

       suggestions_count         platform                    genres  \
0                    416               PC       [Action, Adventure]   
1                    416  Xbox Series S/X       [Action, Adventure]   
2                    416    PlayStation 5       [Action, Adventure]   
3                    416    PlayStation 4       [Action, Adventure]   
4                    416    PlayStation 3       [Action, Adventure]   
...                  ...              ...                       ...   
140611               193          Android          [Action, Puzzle]   
140612                90              iOS  [Action, Arcade, Casual]   
140613                90          Android  [Action, Arcade, Casual]   
140614               365              iOS          [Family, Puzzle]   
140615               365               PC          [Family, Puzzle]   

                                                   stores  \
0       [Epic Games, PlayStation Store, Xbox Store, Xb...   
1       [Epic Games, PlayStation Store, Xbox Store, Xb...   
2       [Epic Games, PlayStation Store, Xbox Store, Xb...   
3       [Epic Games, PlayStation Store, Xbox Store, Xb...   
4       [Epic Games, PlayStation Store, Xbox Store, Xb...   
...                                                   ...   
140611                           [App Store, Google Play]   
140612                           [App Store, Google Play]   
140613                           [App Store, Google Play]   
140614                                          App Store   
140615                                          App Store   

                                                     tags          esrb  
0       [Singleplayer, Steam Achievements, Multiplayer...        Mature  
1       [Singleplayer, Steam Achievements, Multiplayer...        Mature  
2       [Singleplayer, Steam Achievements, Multiplayer...        Mature  
3       [Singleplayer, Steam Achievements, Multiplayer...        Mature  
4       [Singleplayer, Steam Achievements, Multiplayer...        Mature  
...                                                   ...           ...  
140611         [Physics, Cartoon, hero, Monsters, wizard]  Everyone 10+  
140612  [Retro, Robots, character, fun, shoot, art, sp...          Teen  
140613  [Retro, Robots, character, fun, shoot, art, sp...          Teen  
140614  [Cute, Aliens, achievements, Story, fun, cats,...  Everyone 10+  
140615  [Cute, Aliens, achievements, Story, fun, cats,...  Everyone 10+  

[140616 rows x 10 columns]

---
# Checking for mobile games after the unpacking

In [18]:
sorted(rawg_data['platform'].unique())

['3DO',
 'Android',
 'Apple II',
 'Atari 2600',
 'Atari 5200',
 'Atari 7800',
 'Atari 8-bit',
 'Atari Flashback',
 'Atari Lynx',
 'Atari ST',
 'Atari XEGS',
 'Classic Macintosh',
 'Commodore / Amiga',
 'Dreamcast',
 'Game Boy',
 'Game Boy Advance',
 'Game Boy Color',
 'Game Gear',
 'GameCube',
 'Genesis',
 'Jaguar',
 'Linux',
 'NES',
 'Neo Geo',
 'Nintendo 3DS',
 'Nintendo 64',
 'Nintendo DS',
 'Nintendo DSi',
 'Nintendo Switch',
 'PC',
 'PS Vita',
 'PSP',
 'PlayStation',
 'PlayStation 2',
 'PlayStation 3',
 'PlayStation 4',
 'PlayStation 5',
 'SEGA 32X',
 'SEGA CD',
 'SEGA Master System',
 'SEGA Saturn',
 'SNES',
 'Web',
 'Wii',
 'Wii U',
 'Xbox',
 'Xbox 360',
 'Xbox One',
 'Xbox Series S/X',
 'iOS',
 'macOS']

We can see that after unpacking the `platforms` column into `platform`, we get `iOS` and `Android` again.

We are going to drop the raws containing those data by getting the indexes and using the `pd.DataFrame.drop()` method.

In [19]:
ios = list(rawg_data[rawg_data['platform']=='iOS'].index)
android = list(rawg_data[rawg_data['platform']=='Android'].index)

drop = ios + android

rawg_data = rawg_data.drop(index = drop)\
                     .reset_index(drop=True)

rawg_data

slug                         name   released  \
0               grand-theft-auto-v           Grand Theft Auto V 2013-09-17   
1               grand-theft-auto-v           Grand Theft Auto V 2013-09-17   
2               grand-theft-auto-v           Grand Theft Auto V 2013-09-17   
3               grand-theft-auto-v           Grand Theft Auto V 2013-09-17   
4               grand-theft-auto-v           Grand Theft Auto V 2013-09-17   
...                            ...                          ...        ...   
122922        docs-for-playstation         Docs for PlayStation 2016-05-18   
122923        docs-for-playstation         Docs for PlayStation 2016-05-18   
122924                  bee-leader                   Bee Leader 2012-05-24   
122925  delta-strike-first-assault  Delta Strike: First Assault 2016-05-31   
122926            kitten-sanctuary             Kitten Sanctuary 2009-03-13   

        metacritic suggestions_count         platform  \
0             97.0               416               PC   
1             97.0               416  Xbox Series S/X   
2             97.0               416    PlayStation 5   
3             97.0               416    PlayStation 4   
4             97.0               416    PlayStation 3   
...            ...               ...              ...   
122922         NaN                 0    PlayStation 3   
122923         NaN                 0    PlayStation 4   
122924         NaN                45            macOS   
122925         NaN               356          PS Vita   
122926         NaN               365               PC   

                          genres  \
0            [Action, Adventure]   
1            [Action, Adventure]   
2            [Action, Adventure]   
3            [Action, Adventure]   
4            [Action, Adventure]   
...                          ...   
122922                       NaN   
122923                       NaN   
122924  [Action, Arcade, Casual]   
122925                    Action   
122926          [Family, Puzzle]   

                                                   stores  \
0       [Epic Games, PlayStation Store, Xbox Store, Xb...   
1       [Epic Games, PlayStation Store, Xbox Store, Xb...   
2       [Epic Games, PlayStation Store, Xbox Store, Xb...   
3       [Epic Games, PlayStation Store, Xbox Store, Xb...   
4       [Epic Games, PlayStation Store, Xbox Store, Xb...   
...                                                   ...   
122922                                  PlayStation Store   
122923                                  PlayStation Store   
122924                                          App Store   
122925                                  PlayStation Store   
122926                                          App Store   

                                                     tags          esrb  
0       [Singleplayer, Steam Achievements, Multiplayer...        Mature  
1       [Singleplayer, Steam Achievements, Multiplayer...        Mature  
2       [Singleplayer, Steam Achievements, Multiplayer...        Mature  
3       [Singleplayer, Steam Achievements, Multiplayer...        Mature  
4       [Singleplayer, Steam Achievements, Multiplayer...        Mature  
...                                                   ...           ...  
122922                      [Music, online, console, art]           NaN  
122923                      [Music, online, console, art]           NaN  
122924  [Cute, Physics, achievements, Music, city, Fli...           NaN  
122925                     [combat, online, Tanks, drone]  Everyone 10+  
122926  [Cute, Aliens, achievements, Story, fun, cats,...  Everyone 10+  

[122927 rows x 10 columns]

# Getting the url for the metacritic page for each game

## Listing the platforms available in Metacritic into a dictionary

Remember the metacritic page for a game has the next format:

https://www.metacritic.com/game/PLATFORM/SLUG

But not every platform in this dataset is available, as this page did not exist.

We will list the platforms available in Metacritic inside a dictionary, using the `Platform` column counterpart as the key value:
`RAWG-platform: Metacritic-platform`

In [20]:
platforms_metacritic = {
    
    # Sony
    
    "PlayStation 5": "playstation-5",
    "PlayStation 4": "playstation-4",
    "PlayStation 3": "ps3",
    "PlayStation 2": "ps2",
    "PlayStation": "ps",
    "PS Vita": "vita",
    "PSP": "psp",
    
    # Microsoft
    
    "Xbox One": "xbox-one",
    "Xbox Series S/X": "xbox-series-x",
    "Xbox 360": "xbox-360",
    "Xbox": "xbox",
    
    # Nintendo
    
    "Nintendo Switch": "switch",
    "Wii U": "wii-u",
    "Wii": "wii",
    "GameCube": "gamecube",
    "Nintendo 64": "n64",
    "Nintendo 3DS": "3ds",
    "Nintendo DS": "ds",
    "Nintendo DSi": "ds",
    "Game Boy Advance": "gba",
    
    # Others
    
    "PC": "pc",
    "Dreamcast": "dreamcast"
    
}

---

We are going to add a new column `plat_mc` to `rawg_data` for the platform in Metacritic format. Any console not in the `platforms_metacritic` dictionary will have a `NaN` assigned.

In [21]:
def plat_metacritic(x):
    
    if x in platforms_metacritic.keys():
        
        result = platforms_metacritic[x]
        
    else:
        
        result = np.nan
        
    return result

In [22]:
rawg_data['plat_mc'] = rawg_data['platform'].apply(plat_metacritic)

rawg_data

slug                         name   released  \
0               grand-theft-auto-v           Grand Theft Auto V 2013-09-17   
1               grand-theft-auto-v           Grand Theft Auto V 2013-09-17   
2               grand-theft-auto-v           Grand Theft Auto V 2013-09-17   
3               grand-theft-auto-v           Grand Theft Auto V 2013-09-17   
4               grand-theft-auto-v           Grand Theft Auto V 2013-09-17   
...                            ...                          ...        ...   
122922        docs-for-playstation         Docs for PlayStation 2016-05-18   
122923        docs-for-playstation         Docs for PlayStation 2016-05-18   
122924                  bee-leader                   Bee Leader 2012-05-24   
122925  delta-strike-first-assault  Delta Strike: First Assault 2016-05-31   
122926            kitten-sanctuary             Kitten Sanctuary 2009-03-13   

        metacritic suggestions_count         platform  \
0             97.0               416               PC   
1             97.0               416  Xbox Series S/X   
2             97.0               416    PlayStation 5   
3             97.0               416    PlayStation 4   
4             97.0               416    PlayStation 3   
...            ...               ...              ...   
122922         NaN                 0    PlayStation 3   
122923         NaN                 0    PlayStation 4   
122924         NaN                45            macOS   
122925         NaN               356          PS Vita   
122926         NaN               365               PC   

                          genres  \
0            [Action, Adventure]   
1            [Action, Adventure]   
2            [Action, Adventure]   
3            [Action, Adventure]   
4            [Action, Adventure]   
...                          ...   
122922                       NaN   
122923                       NaN   
122924  [Action, Arcade, Casual]   
122925                    Action   
122926          [Family, Puzzle]   

                                                   stores  \
0       [Epic Games, PlayStation Store, Xbox Store, Xb...   
1       [Epic Games, PlayStation Store, Xbox Store, Xb...   
2       [Epic Games, PlayStation Store, Xbox Store, Xb...   
3       [Epic Games, PlayStation Store, Xbox Store, Xb...   
4       [Epic Games, PlayStation Store, Xbox Store, Xb...   
...                                                   ...   
122922                                  PlayStation Store   
122923                                  PlayStation Store   
122924                                          App Store   
122925                                  PlayStation Store   
122926                                          App Store   

                                                     tags          esrb  \
0       [Singleplayer, Steam Achievements, Multiplayer...        Mature   
1       [Singleplayer, Steam Achievements, Multiplayer...        Mature   
2       [Singleplayer, Steam Achievements, Multiplayer...        Mature   
3       [Singleplayer, Steam Achievements, Multiplayer...        Mature   
4       [Singleplayer, Steam Achievements, Multiplayer...        Mature   
...                                                   ...           ...   
122922                      [Music, online, console, art]           NaN   
122923                      [Music, online, console, art]           NaN   
122924  [Cute, Physics, achievements, Music, city, Fli...           NaN   
122925                     [combat, online, Tanks, drone]  Everyone 10+   
122926  [Cute, Aliens, achievements, Story, fun, cats,...  Everyone 10+   

              plat_mc  
0                  pc  
1       xbox-series-x  
2       playstation-5  
3       playstation-4  
4                 ps3  
...               ...  
122922            ps3  
122923  playstation-4  
122924            NaN  
122925           vita  
122926             pc  

[122927 rows x 11 columns]

In [23]:
rawg_data['plat_mc'].value_counts()

pc               57716
playstation-4     5495
switch            4193
xbox-one          4149
ps3               3485
xbox-360          2476
wii               2283
ds                2182
vita              1895
3ds               1691
psp               1546
ps2               1513
ps                1356
wii-u             1251
gba                854
xbox               671
gamecube           625
n64                338
dreamcast          324
playstation-5      117
xbox-series-x       96
Name: plat_mc, dtype: int64

---

Indexes seem to have set all to 0 somehow, so we will reset them with the `pd.DataFrame.reset_index()` method.

In [24]:
rawg_data.head()

slug                name   released  metacritic  \
0  grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   
1  grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   
2  grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   
3  grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   
4  grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   

  suggestions_count         platform               genres  \
0               416               PC  [Action, Adventure]   
1               416  Xbox Series S/X  [Action, Adventure]   
2               416    PlayStation 5  [Action, Adventure]   
3               416    PlayStation 4  [Action, Adventure]   
4               416    PlayStation 3  [Action, Adventure]   

                                              stores  \
0  [Epic Games, PlayStation Store, Xbox Store, Xb...   
1  [Epic Games, PlayStation Store, Xbox Store, Xb...   
2  [Epic Games, PlayStation Store, Xbox Store, Xb...   
3  [Epic Games, PlayStation Store, Xbox Store, Xb...   
4  [Epic Games, PlayStation Store, Xbox Store, Xb...   

                                                tags    esrb        plat_mc  
0  [Singleplayer, Steam Achievements, Multiplayer...  Mature             pc  
1  [Singleplayer, Steam Achievements, Multiplayer...  Mature  xbox-series-x  
2  [Singleplayer, Steam Achievements, Multiplayer...  Mature  playstation-5  
3  [Singleplayer, Steam Achievements, Multiplayer...  Mature  playstation-4  
4  [Singleplayer, Steam Achievements, Multiplayer...  Mature            ps3

In [25]:
rawg_data = rawg_data.reset_index(drop=True)

We will list the indexes of `NaN` in the `plat_mc` column in order to drop them.

In [26]:
no_metacritic = list(rawg_data[rawg_data['plat_mc'].isna()].index)

In [27]:
rawg_data = rawg_data.drop(index = no_metacritic)\
                     .reset_index(drop = True)

---
We are creating a new column named `url`, which will include the hyperlink to the games' Metacritic page (if it exists). This will ease the web scraping process.

In [28]:
rawg_data['url'] = 'https://www.metacritic.com/game/' + rawg_data['plat_mc'] + '/' + rawg_data['slug']

rawg_data.head()

slug                name   released  metacritic  \
0  grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   
1  grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   
2  grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   
3  grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   
4  grand-theft-auto-v  Grand Theft Auto V 2013-09-17        97.0   

  suggestions_count         platform               genres  \
0               416               PC  [Action, Adventure]   
1               416  Xbox Series S/X  [Action, Adventure]   
2               416    PlayStation 5  [Action, Adventure]   
3               416    PlayStation 4  [Action, Adventure]   
4               416    PlayStation 3  [Action, Adventure]   

                                              stores  \
0  [Epic Games, PlayStation Store, Xbox Store, Xb...   
1  [Epic Games, PlayStation Store, Xbox Store, Xb...   
2  [Epic Games, PlayStation Store, Xbox Store, Xb...   
3  [Epic Games, PlayStation Store, Xbox Store, Xb...   
4  [Epic Games, PlayStation Store, Xbox Store, Xb...   

                                                tags    esrb        plat_mc  \
0  [Singleplayer, Steam Achievements, Multiplayer...  Mature             pc   
1  [Singleplayer, Steam Achievements, Multiplayer...  Mature  xbox-series-x   
2  [Singleplayer, Steam Achievements, Multiplayer...  Mature  playstation-5   
3  [Singleplayer, Steam Achievements, Multiplayer...  Mature  playstation-4   
4  [Singleplayer, Steam Achievements, Multiplayer...  Mature            ps3   

                                                 url  
0  https://www.metacritic.com/game/pc/grand-theft...  
1  https://www.metacritic.com/game/xbox-series-x/...  
2  https://www.metacritic.com/game/playstation-5/...  
3  https://www.metacritic.com/game/playstation-4/...  
4  https://www.metacritic.com/game/ps3/grand-thef...

We save `rawg_data` to a `.csv` file for its use after in a new notebook to ease the reading.

In [29]:
rawg_data.to_csv("02_rawg_metacritic_url.csv", encoding='utf-8', index=False)